crates/typed-store/src/sally/mod.rs

```diff
@@ -0,0 +1,588 @@
+//!     update_table(&table);
+//!     Ok(())
+//! }
+//! ```
+use crate::{
+    rocks::{
+        default_db_options, keys::Keys, values::Values, DBBatch, DBMap, DBOptions,
+        RocksDBAccessType, TypedStoreError,
+    },
+    test_db::{TestDB, TestDBIter, TestDBKeys, TestDBValues, TestDBWriteBatch},
+    traits::Map,
+};
+
+use crate::rocks::iter::Iter as RocksDBIter;
+use crate::rocks::DBMapTableConfigMap;
+use collectable::TryExtend;
+use rocksdb::Options;
+use serde::{de::DeserializeOwned, Serialize};
+use std::borrow::Borrow;
+use std::{collections::BTreeMap, path::PathBuf};
+
+pub enum SallyRunMode {
+    // Whether Sally should use its own memtable and wal for read/write or just fallback to
+    // reading/writing directly from the backend db. When columns in the db are backed by different
+    // backend stores, we should never use `FallbackToDB` as that would lose atomicity,
+    // transactions and db recovery
+    FallbackToDB,
+}
+
+pub struct SallyConfig {
+    pub mode: SallyRunMode,
+}
+
+impl Default for SallyConfig {
+    fn default() -> Self {
+        Self {
+            mode: SallyRunMode::FallbackToDB,
+        }
+    }
+}
+
+/// A Sally column could be anything that implements key value interface. We will eventually have
+/// Sally serve read/writes using its own memtable and wal when columns in the db are backend by more then
+/// one backend store (e.g different rocksdb instances and/or distributed key value stores)
+pub enum SallyColumn<K, V> {
+    RocksDB((DBMap<K, V>, SallyConfig)),
+    TestDB((TestDB<K, V>, SallyConfig)),
+}
+
+impl<K, V> SallyColumn<K, V> {
+    pub fn new_single_rocksdb(db: DBMap<K, V>) -> Self {
+        // When all columns in the db are backed by a single rocksdb instance, we will fallback to
+        // using native rocksdb read and write apis and use default config
+        SallyColumn::RocksDB((db, SallyConfig::default()))
+    }
+    pub fn new_testdb(db: TestDB<K, V>) -> Self {
+        SallyColumn::TestDB((db, SallyConfig::default()))
+    }
+    pub fn batch(&self) -> SallyWriteBatch {
+        match self {
+            SallyColumn::RocksDB((
+                db_map,
+                SallyConfig {
+                    mode: SallyRunMode::FallbackToDB,
+                },
+            )) => SallyWriteBatch::RocksDB(db_map.batch()),
+            SallyColumn::TestDB((
+                test_db,
+                SallyConfig {
+                    mode: SallyRunMode::FallbackToDB,
+                },
+            )) => SallyWriteBatch::TestDB(test_db.batch()),
+        }
+    }
+}
+
+impl<'a, K, V> Map<'a, K, V> for SallyColumn<K, V>
+where
+    K: Serialize + DeserializeOwned,
+    V: Serialize + DeserializeOwned,
+{
+    type Error = TypedStoreError;
+    type Iterator = SallyIter<'a, K, V>;
+    type Keys = SallyKeys<'a, K>;
+    type Values = SallyValues<'a, V>;
+
+    fn contains_key(&self, key: &K) -> Result<bool, TypedStoreError> {
```

Comment:
i think all the read methods (at least) should be async - first because we may wish to use an async backend, second because we might later want to aggregate independent reads into multi-reads for efficiency.

Reply:
Agree with this, i'll make this change